utils/message/spoiler.go

```diff
@@ -0,0 +1,154 @@
+	contentStartIndex := 0
+
+	for i := 0; i < msgLen; i++ {
+		r := msgRunes[i]
+		switch r {
+		case '!':
+			// 連続する!をトークンに変換
+			c := countPrefixRune(msgRunes[i:], '!')
+			if c >= 2 {
+				if i != contentStartIndex {
+					result = append(result, spoilerToken{tType: spoilerTokenContent, body: msgRunes[contentStartIndex:i]})
+				}
+
+				for j := 0; j < c/2; j++ {
+					result = append(result, spoilerToken{tType: spoilerTokenExclamation, body: msgRunes[i : i+2]})
+				}
+				i += c - 1
+				if c%2 == 1 { // !が奇数個だった場合最後のものは処理していない
+					i--
+				}
+				contentStartIndex = i + 1
+			}
+		case '\r', '\n', ' ', '　':
+			if i != contentStartIndex {
+				result = append(result, spoilerToken{tType: spoilerTokenContent, body: msgRunes[contentStartIndex:i]})
+			}
+			result = append(result, spoilerToken{tType: spoilerTokenSplit, body: msgRunes[i : i+1]})
+			contentStartIndex = i + 1
+		}
+	}
+
+	if msgLen != contentStartIndex {
+		result = append(result, spoilerToken{tType: spoilerTokenContent, body: msgRunes[contentStartIndex:msgLen]})
+	}
+	return result
+}
+
+var emptyRuneSlice = []rune{}
+
+func tokensToString(tokens []spoilerToken) string {
+	spoilerStartPos := []int{}
+	spoilerEndPos := []int{}
+
+	tokensLen := len(tokens)
+	for i, current := range tokens {
+		var prev spoilerToken
+		var next spoilerToken
+		if i > 0 {
+			prev = tokens[i-1]
+		}
+		if i+1 < tokensLen-1 {
+			next = tokens[i+1]
+		}
+
+		if current.tType == spoilerTokenExclamation {
+			if len(spoilerStartPos) > len(spoilerEndPos) {
+				if prev.tType != spoilerTokenInvalid &&
+					prev.tType != spoilerTokenSplit &&
+					spoilerStartPos[len(spoilerStartPos)-1] != i-1 {
+					// 閉じれたら閉じる
+					spoilerEndPos = append(spoilerEndPos, i)
+				} else if next.tType != spoilerTokenSplit {
+					// 閉じれなくても開けたら開く
+					spoilerStartPos = append(spoilerStartPos, i)
+				}
+			} else {
+				if next.tType != spoilerTokenInvalid && next.tType != spoilerTokenSplit {
+					spoilerStartPos = append(spoilerStartPos, i)
+				}
+			}
+		}
+	}
+
+	// 個数があっていないときは対応関係を正す
+	if len(spoilerStartPos) > len(spoilerEndPos) {
+		newSpoilerStartPos := make([]int, 0, len(spoilerStartPos))
+		readEndCount := 0
+		spoilerEndPosLen := len(spoilerEndPos)
+		for i := len(spoilerStartPos) - 1; i >= 0 && readEndCount < spoilerEndPosLen; i-- {
+			start := spoilerStartPos[i]
+			end := spoilerEndPos[spoilerEndPosLen-1-readEndCount]
+			if end < start {
+				continue
+			}
+			newSpoilerStartPos = append(newSpoilerStartPos, start)
+			readEndCount++
+		}
+
+		// newSpoilerStartPosの順番を逆転
+		for i := 0; i < len(newSpoilerStartPos)/2; i++ {
+			newSpoilerStartPos[i], newSpoilerStartPos[len(newSpoilerStartPos)-i-1] = newSpoilerStartPos[len(newSpoilerStartPos)-i-1], newSpoilerStartPos[i]
+		}
+		spoilerStartPos = newSpoilerStartPos
+	}
+
+	for i := 0; i < len(spoilerStartPos); i++ {
+		s := spoilerStartPos[i]
+		e := spoilerEndPos[i]
+		tokens[s].body = emptyRuneSlice
+		tokens[e].body = emptyRuneSlice
```

Comment:
グローバルにemptyRuneSliceをvarで置くのではなく、nil (or empty sliceを生成)の方がわかりやすいかも？
```suggestion
		tokens[s].body = nil
		tokens[e].body = nil
```